pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -313,3 +331,69 @@ func (c *Repair) checkLeaked(leaks map[string]int, stored ipallocator.Interface,
 		}
 	})
 }
+
+// used to generate empty patch bytes
+type emptyType struct {
+}
+
+// Services with "PreferDualStack" would be
+// -- set to dual-stack on a dual-stack cluster.
+// -- set to single stack on a single stack cluster.
+// Because the way allocator work (on Create/Update) they won't be upgraded
+// downgraded (from/to dual-stack) as cluster config change. More over
+// they will be arbitrary upgraded/downgraded if the user tried to
+// update them (e.g. setting a label).
+// This func ensures that all services of policy "PreferDualStack" is aligned
+// to cluster config without the user needing to update them
+func (c *Repair) fixPreferDualStackServices() error {
+	isClusterDualStack := len(c.allocatorByFamily) == 2
+
+	svcList, err := c.serviceClient.Services(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
```

Comment:
do we need to list and go through all the services twice now?
https://github.com/kubernetes/kubernetes/pull/101914/commits/5480b8ba5dd35bb3315eee709788096108f35b54#diff-4b26a9664c2321f036fc331bad12530015616fd933aa9c66300ec86715347121R211